techlibs/quicklogic/synth_quicklogic.cc

```diff
@@ -0,0 +1,162 @@
+        log("        is omitted if this parameter is not specified.\n");
+        log("\n");
+        log("    -blif <file>\n");
+        log("        write the design to the specified BLIF file. writing of an output file\n");
+        log("        is omitted if this parameter is not specified.\n");
+        log("\n");
+        log("    -flatten\n");
+        log("        flatten design before synthesis\n");
+        log("\n");
+        help_script();
+        log("\n");
+    }
+
+    std::string top_opt = "-auto-top";
+    std::string edif_file = "";
+    std::string blif_file = "";
+    std::string currmodule = "";
+    bool flatten = false;
+
+    void clear_flags() YS_OVERRIDE
+    {
+        top_opt = "-auto-top";
+        flatten = false;
+        edif_file.clear();
+        blif_file.clear();
+    }
+
+    void execute(std::vector<std::string> args, RTLIL::Design *design) YS_OVERRIDE
+    {
+        std::string run_from, run_to;
+        clear_flags();
+
+        size_t argidx;
+        for (argidx = 1; argidx < args.size(); argidx++)
+        {
+            if (args[argidx] == "-top" && argidx+1 < args.size()) {
+                top_opt = "-top " + args[++argidx];
+                continue;
+            }
+            if (args[argidx] == "-edif" && argidx+1 < args.size()) {
+                edif_file = args[++argidx];
+                continue;
+            }
+            if (args[argidx] == "-blif" && argidx+1 < args.size()) {
+                blif_file = args[++argidx];
+                continue;
+            }
+            if (args[argidx] == "-flatten") {
+                flatten = true;
+                continue;
+            }
+            break;
+        }
+        extra_args(args, argidx, design);
+
+        if (!design->full_selection())
+            log_cmd_error("This command only operates on fully selected designs!\n");
+
+        log_header(design, "Executing SYNTH_QUICKLOGIC pass.\n");
+        log_push();
+
+        run_script(design, run_from, run_to);
+
+        log_pop();
+    }
+
+    void script() YS_OVERRIDE
+    {
+        if (check_label("begin")) {
+            run("read_verilog -lib +/quicklogic/cells_sim.v");
+            run(stringf("hierarchy -check %s", top_opt.c_str()));
+        }
+
+        if (check_label("prepare")) {
+            run("proc");
+            if (flatten || help_mode)
+                run("flatten", "(with '-flatten')");
+        }
+
+        if (check_label("coarse")) {
+            run("tribuf -logic");
+            run("opt_expr");
+            run("opt_clean");
+            run("check");
+            run("opt");
+            run("peepopt");
+            run("opt_clean");
+            run("techmap");
+            run("abc -lut 1:4");
+            run("opt_clean");
+        }
+
+        if (check_label("map")) {
+            run("techmap -map +/quicklogic/cells_map.v");
+            run("clean");
+            run("check");
+            run("opt_clean -purge");
```

Comment:
Are you sure that the `-purge` is needed here? This will throw away all alias names for wires but one, which might make clock constraints harder as you will have to guess the name Yosys chooses.

Reply:
It turned out that all `-purge` were redundant / not necessary. Removed.